File: backend/agent/node-registry.ts
/**
 * Node Registry — classifica e rotula todos os tipos de nodes que o LangGraph pode emitir.
 *
 * O LangGraph emite updates para vários tipos de nodes. Este módulo centraliza
 * a lógica de categorização para que o normalizer e o frontend saibam o que exibir.
 */

export enum NodeCategory {
  /** Invocação direta do LLM (agent, model) */
  LLM_INVOKE = "LLM_INVOKE",
  /** Execução de tools (tools, tool_executor) */
  TOOL_EXECUTION = "TOOL_EXECUTION",
  /** Subgraph de outro agente (formato "agentName:nodeName") */
  SUBGRAPH = "SUBGRAPH",
  /** Nó interno/middleware — não deve aparecer no frontend */
  INTERNAL = "INTERNAL",
  /** Nó customizado desconhecido */
  UNKNOWN = "UNKNOWN",
}

/** Nomes de nós internos que nunca devem ser exibidos */
const INTERNAL_NODE_PATTERNS: RegExp[] = [
  /^__/,                          // __start__, __end__, __interrupt__
  /Middleware/i,                  // patchToolCallsMiddleware, SummarizationMiddleware
  /\.before_/,                    // .before_agent
  /\.after_/,                     // .after_agent
  /^model_request$/,
  /^model_response$/,
  /^patchToolCalls/,
];

/** Nomes canônicos de nodes LLM */
const LLM_NODES = new Set(["agent", "model", "llm", "generate", "chat"]);

/** Nomes canônicos de nodes de tools */
const TOOL_NODES = new Set(["tools", "tool_executor", "tool_node", "action"]);

export function classifyNode(nodeName: string): NodeCategory {
  if (!nodeName) return NodeCategory.INTERNAL;

  // Subgraph: contém ":" separando agente:nó
  if (nodeName.includes(":")) return NodeCategory.SUBGRAPH;

  // Interno
  for (const pattern of INTERNAL_NODE_PATTERNS) {
    if (pattern.test(nodeName)) return NodeCategory.INTERNAL;
  }

  if (LLM_NODES.has(nodeName)) return NodeCategory.LLM_INVOKE;
  if (TOOL_NODES.has(nodeName)) return NodeCategory.TOOL_EXECUTION;

  return NodeCategory.UNKNOWN;
}

export function getNodeLabel(nodeName: string): string {
  if (!nodeName) return "Node";

  // Subgraph: "coder:agent" → "Coder › Agent"
  if (nodeName.includes(":")) {
    const [parent, child] = nodeName.split(":", 2);
    return `${titleCase(parent)} › ${titleCase(child)}`;
  }

  const canonical: Record<string, string> = {
    agent: "Agent",
    tools: "Tools",
    model: "Model",
    llm: "LLM",
    generate: "Generate",
    chat: "Chat",
    tool_executor: "Tools",
    tool_node: "Tools",
    action: "Action",
  };

  return canonical[nodeName] ?? titleCase(nodeName);
}

/** Retorna true se este node deve ter seus eventos expostos ao frontend */
export function isStreamableNode(nodeName: string): boolean {
  const category = classifyNode(nodeName);
  return (
    category === NodeCategory.LLM_INVOKE ||
    category === NodeCategory.TOOL_EXECUTION ||
    category === NodeCategory.SUBGRAPH ||
    category === NodeCategory.UNKNOWN
  );
}

function titleCase(value: string): string {
  if (!value) return "";
  return value
    .replace(/[_-]+/g, " ")
    .trim()
    .replace(/\s+/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase());
}

File: python/omnimind_backend/runtime/registry/node_registry.py
import re
from enum import StrEnum


class NodeCategory(StrEnum):
    LLM_INVOKE = "LLM_INVOKE"
    TOOL_EXECUTION = "TOOL_EXECUTION"
    SUBGRAPH = "SUBGRAPH"
    INTERNAL = "INTERNAL"
    UNKNOWN = "UNKNOWN"


_INTERNAL_PATTERNS = [
    re.compile(r"^__"),
    re.compile(r"Middleware", re.IGNORECASE),
    re.compile(r"\.before_"),
    re.compile(r"\.after_"),
    re.compile(r"^model_request$"),
    re.compile(r"^model_response$"),
    re.compile(r"^patchToolCalls"),
]

_LLM_NODES = {"agent", "model", "llm", "generate", "chat"}
_TOOL_NODES = {"tools", "tool_executor", "tool_node", "action"}


def classify_node(node_name: str) -> NodeCategory:
    if not node_name:
        return NodeCategory.INTERNAL
    if ":" in node_name:
        return NodeCategory.SUBGRAPH

    for pattern in _INTERNAL_PATTERNS:
        if pattern.search(node_name):
            return NodeCategory.INTERNAL

    if node_name in _LLM_NODES:
        return NodeCategory.LLM_INVOKE
    if node_name in _TOOL_NODES:
        return NodeCategory.TOOL_EXECUTION
    return NodeCategory.UNKNOWN


def is_streamable_node(node_name: str) -> bool:
    category = classify_node(node_name)
    return category in {
        NodeCategory.LLM_INVOKE,
        NodeCategory.TOOL_EXECUTION,
        NodeCategory.SUBGRAPH,
        NodeCategory.UNKNOWN,
    }


def get_node_label(node_name: str) -> str:
    if not node_name:
        return "Node"

    if ":" in node_name:
        parent, child = node_name.split(":", 1)
        return f"{_title_case(parent)} › {_title_case(child)}"

    canonical = {
        "agent": "Agent",
        "tools": "Tools",
        "model": "Model",
        "llm": "LLM",
        "generate": "Generate",
        "chat": "Chat",
        "tool_executor": "Tools",
        "tool_node": "Tools",
        "action": "Action",
    }
    return canonical.get(node_name, _title_case(node_name))


def _title_case(value: str) -> str:
    words = value.replace("_", " ").replace("-", " ").strip().split()
    return " ".join(word.capitalize() for word in words)
